File: backend/dataset/rotate/rotate_9.py
import math

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
		
		# Remove redundant passes
        shift = k % n
        gcd = math.gcd(shift, n)
        
        for j in range(gcd):
            i, temp = shift + j, nums[j]
            while i != j:
                nums[i], temp, i = temp, nums[i], (i + shift) % n

            nums[i] = temp
        

File: backend/CDataset/rotate/rotate_9.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
        if(k>n)
        {
            k = k%n;
        }
        int i=0,j=n-k-1;
        while(i<j)
        {
            swap(nums[i],nums[j]);
            i++;
            j--;
        }
        i=n-k,j=n-1;
        while(i<j)
        {
            swap(nums[i],nums[j]);
            i++;
            j--;
        }
        i=0,j=n-1;
        while(i<j)
        {
            swap(nums[i],nums[j]);
            i++;
            j--;
        }
        // return nums;
    }
};
